### PR TITLE
Prevent user to be marked as deleted when LDAP connection failed

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !28486

### CODE REVIEW

File: src/Auth.php [hunk 1]
```diff
@@ -927,7 +943,8 @@ public function login($login_name, $login_password, $noauto = false, $remember_m
                                 $login_password,
                                 $this->user->fields["auths_id"]
                             );
-                            if ($this->ldap_connection !== false && (!$this->auth_succeded && !$this->user_found)) {
+                            if ($this->user_ldap_error === false && (!$this->auth_succeded && !$this->user_found)) {
```
Comment (cedric-anne) [inline src/Auth.php:946]: Main problem was located here. As long as the connection to the last LDAP server of the list was successful (i.e. bind with root DN was OK), `$this->ldap_connection` was true.
Following logic was resulting in user deletion:
 - try to connect to server where user was seen on last login/sync: connection failure => `$this->ldap_connection = false; $this->user_found = false;`;
 - try to connect to another server: connection OK but user not found => `$this->ldap_connection = \LDAP\Connection; $this->user_found = false;`;
 - result : `$this->ldap_connection !== false && (!$this->auth_succeded && !$this->user_found)` was `true`.

**Comment by cedric-anne** [comment]
> Not yet fixed, I have to add some more tests.

It should be OK now.

**Review by flegastelois — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Comment by cedric-anne** [comment]
I figured out that after removing the `error_reporting(0)` that was surrounding the LDAP auth, bind errors may be added in logs when user typed a wrong login/pass. I added a commit to fix this.

Also, In `AuthLDAP::tryToConnectToServer()`, I added a small fix to prevent a useless call to `AuthLDAP::connectToServer()` with user login/pass when `$ldap_method['use_bind']` is `false`. Indeed, it would produce the same result as call made a few line above.

File: src/Auth.php [hunk 2]
```diff
@@ -271,19 +280,26 @@ public function connection_ldap($ldap_method, $login, $password)
                 ]);
             } catch (\Throwable $e) {
                 ErrorHandler::getInstance()->handleException($e, true);
+                $info = false;
+            }
+
+            $ldap_errno = ldap_errno($this->ldap_connection);
+            if ($info === false && $ldap_errno > 0 && $ldap_errno !== 32) {
+                // 32 = LDAP_NO_SUCH_OBJECT => This should not be considered as a connection error, as it just means that user was not found.
```
Comment (trasher) [inline src/Auth.php:288]: Indeed, this is not a connection error, but we know at this point this is already NOK.
I would have expected the code to stop properly here if user is not found; am I wrong?

**Review by trasher — APPROVED** [review]
LGTM

**Review by orthagh — APPROVED** [review]

**Pull request merged**